Where.not and nil (#45)
andycroll/andycroll.com: _posts/2019/2019-04-28-be-aware-of-nil-values-when-using-where-not.md

---
title: Be Aware of nil values when using where.not()
description: Strange Think
layout: article
category: ruby
image:
  base: '2019/be-aware-of-nil-values-when-using-where-not'
  alt: 'Half-finished painted wall'
  source: "https://unsplash.com/photos/oBb-Y26PJgg"
  credit: "Samuel Zeller"
---

The use of `where.not` when building Active Record scopes can help you to elegantly specify scopes for cases where an attribute does not have a specific value.

> However the abstraction has some quirks if the attribute you’re querying can be `NULL` in your database (`nil` in Ruby). The `where.not` scope doesn’t automatically return `nil` values unless you specify them.

Reviewer: *unless you specify that you want them (i.e. want them returned)?

## Instead of…

…expecting `where.not` to return `nil` values:

```ruby
non_oat_options = Coffee.where.not(milk: "Oatly")
```

## Use…

…an `or` clause to explicitly request the `nil` values.

```ruby
non_oat_options = Coffee.where.not(milk: "Oatly").or(Coffee.where(milk: nil))
```

## But why?

Of course this all depends on whether you are expecting the answer to include those nil values!

> The standard usage `where(milk: "Oatly")`, as expected, does not include `nil` values. However in Rubyit looks as though `where.not(milk: "Oatly")` would be the inverse: _but it isn’t_.

Reviewer: *Ruby it

Reviewer: ```suggestion
`where(milk: "Oatly")`, as expected, will not return `nil` values, and you would think that `where.not(milk: "Oatly")` would be the complete inverse: _but it isn’t_.
```

Not sure if 'complete' makes sense here, but it does to me...to get across: 'everything else' not returned in `where(milk: "Oatly")`.

> This is because the SQL result of a `where.not` scope is SQL’s `!=` operator which doesn't return `NULL` values from the database.

Reviewer: * comma before 'which'.
